### ISSUE TITLE
[ci skip] for pull request

### ISSUE THREAD

**Issue by mathieujobin** [travis-ci/travis-ci#5374]
`[ci skip]` is very useful per commit. but sometimes you want to open a pull request early for review but you are not necessarely ready to run the tests right away

if I could add [ci skip] to my PR title it would be perfect
when I am ready, I am simply taking it out. I might need to close and reopen the PR but I am fine with that. I am already doing it as a trick to rerun the pull request against updated master ;)

cheers,

**Comment by BanzaiMan** [comment]
How do you expect maintainers of the upstream project to evaluate the PR without actually running the build?

**Comment by Joshua-Anderson** [comment]
This could be useful for documentation PRs. Many projects have required commit formats, so it's not possible to add a `[ci skip]` to the commit itself.

**Comment by mathieujobin** [comment]
the idea is to have a code review from your peers, before it is ready to be sent to the CI server, especially for projects with a lot of tests and especially because every commits trigger two build.

once the submitter is ready to have the test being run by travis, he can edit the PR title to remove the `[ci skip]` tag and click close/reopen the PR to trigger the build

I'm not sure If I am describing this better this time ?

**Comment by kbrock** [comment]
> @BanzaiMan
>How do you expect maintainers of the upstream project to evaluate the PR without actually running the build?

Currently, a person submitting a PR is able to skip a build by putting `[ci skip]` into all the commits. So allowing them to put that same phrase into the PR text is granting them no more or less power.

This is nice because the maintainers can edit the title, remove `[ci skip]`, and trigger a build. If that text were in the commits themselves, only the submitter can trigger a build, not the maintainer.

**Comment by joshk** [comment]
I'm sorry but we don't check the PR title for testability.

I do understand the point, and how this might be useful.

We will have a think about this, but I do not have an ETA on if or when we will make a decision.

**Comment by cmw** [comment]
@joshk We have a use case where this is useful, as we start PRs in our github immediately to have access to the comment section for discussion purposes. We currently flag those [WIP] so we know not to care about CI status etc yet, but pushes to those PRs keep clogging up our travis workers. A way to skip those branches would be great for our deploy pipeline.

**Comment by rameshthoomu** [comment]
@joshk : I have a scenario where i want to skip the PR for document change or for small code correction where the build deployment and further tests are not required. Could you please give me a solution how to skip when some sends a PR against the Document changes. thanks in advance!!

**Comment by se7entyse7en** [comment]
@joshk I have a similar use case to that exposed by @cmw

**Comment by sbts** [comment]
We also have a similar use case, updates to scripts and docs that don't affect anything in the testable code shouldn't need to trigger a travis build and test. (our main project is a perl project where we use travis to run our test suite, but there is nothing for travis to actually build.

**Comment by akien-mga** [comment]
I'd also like to have this feature, especially for documentation updates as other mentioned, or when adding icons, moving scripts around... basically anything that does not impact the buildsystem.

I like to keep my git history clean, and adding `[ci skip]` to all such commit logs is far from optimal; 2 years from now, when I read the git history, I don't really need to know that the CI was skipped for this specific commit. That's why I'd like to be able to keep this kind of metadata in GitHub's PR system, so that I can get the feature without touching the git history.

**Comment by ehuelsmann** [comment]
Same here.

**Comment by jakirkham** [comment]
So I actually strongly dislike this idea and would like to elucidate why.

One of the use cases one of my orgs has is running CI on the PR, but not on the merge commit. In these cases, what we want to do is have the title include `[ci skip]` as that will get included in the merge commit automatically when merging. This way the commits in the PR can be verified to be working before the merge. Then after merging we simply skip the Travis CI build.

Having titles included as part of CI build determination for the PR would actually break our use case.

**Comment by mathieujobin** [comment]
@jakirkham skipping merge commit on master is a wrong idea.
even if the PR build passed, unless you force your branch to be up to date before merge, you are taking a chance to introduce incompatibilities between merges

**Comment by jakirkham** [comment]
Not really @mathieujobin as Travis CI tests the merge commit that GitHub creates by "merging" with the base branch. The quotes are not to suggest it isn't a real merge, but merely to suggest the PR itself is still not yet merged.

In general, if you are concerned about the testing being stale (e.g. other things got merged first), then yes you will either need to close/reopen the PR to make GitHub generate a new merge ref or you need to update your PR by merging/rebasing with the base branch. However, I don't see this as a fundamental problem of my use case. Instead this is a problem with merging PRs with stale CI results regardless of the reason. However, if you do have truly orthogonal unit tests, this is very rarely any practical concern.

**Comment by cmw** [comment]
@jakirkham So you basically strongly dislike this idea on account that "our" usecase would break your usecase. I'd actually advocate for having two different tags for ignoring branch builds and commits. Say, [WIP] and [ci skip] (the former as part of the branch name being clearly transient, the latter being part of a commit message that will probably not change).

**Comment by kbrock** [comment]
Every community does something different. @mathieujobin likes running CI on PRs, @jakirkham likes running CI after merging with master, and my team likes running CI on both. Kudos to Travis for giving the power to the community to decide for themselves.

Some communities want to use `[ci skip]` in commits, others want to use it in PRs, and probably others want to never use it. Just let each community decides when it is appropriate to use it.

@jakirkham It sounds like this feature will not work for your community. Is it possible to simply ask the users of your community to not use it? Or, maybe remove it from the title of PRs that have it?

@cmw It would be nice if there weren't too many variations to remember. So I'd prefer just keeping it to `[ci skip]`. I thought you were going to propose configuring the phrase that will cause travis to skip.

**Comment by jakirkham** [comment]
> Every community does something different. @mathieujobin likes running CI on PRs, @jakirkham likes running CI after merging with master, and my team likes running CI on both. Kudos to Travis for giving the power to the community to decide for themselves.

👍

> It sounds like this feature will not work for your community. Is it possible to simply ask the users of your community to not use it? Or, maybe remove it from the title of PRs that have it?

Not really. It will overwhelm the CIs. It also breaks with non-Travis CIs, which makes life a bit more problematic.

TBH I don't see why it has to be all one way or another. My objection is simply that your feature is activated by using `[ci skip]` or `[skip ci]` and thus breaks an existing use case.

Why not use a different name like `[ci skip all]` or `[skip all ci]`?

**Comment by kbrock** [comment]
@jakirkham ok, I just picked up on your nuance. Skipping ci on the merge into master is done because the merge commit has `[skip ci]` in it - not because you asked travis to simply not test merges into master. So I take it that you sometimes want to run CI on master merges, and other times want to skip it? I had thought that Travis allowed you to blanket skip CI for merges into master.

As you already said, if you put this phrase in the PR title, it will not run CI in the PR. When that was merged into master, it would retain that phrase, and not run there either. Which is not good.

In summary:
You want the skip phrase in the PR title (that says skip CI for this PR) to be different from the skip phrase in the commit message (that says skip CI for this merge into master).

**Comment by jakirkham** [comment]
> So I take it that you sometimes want to run CI on master merges, and other times want to skip it?

Correct. Only sometimes we want to do this.

**Comment by cmw** [comment]
Having this as the same default for both PR and commit and both individually configurable would be perfectly fine IMO

**Comment by mathieujobin** [comment]
> Correct. Only sometimes we want to do this.

@jakirkham this feature won't even break your use case.

you simply have to not use the `[ci skip]` while the PR is open, being reviewed and evolving into the most beautiful piece of art... making sure tests are passing and all. 

Then just before merging, you rename the PR to include that flag, hit merge. and you won't get that extra build on travis.

it is the exact opposite of my use case, where I want to use this `[ci skip]` tag while the PR is in early review and I don't want to trigger a hour long build when I know I am not done. Then, once I am ready, I take it out of the PR title. and things resume as ordinary.

See everyone is happy, thank you Travis !

**Comment by kbrock** [comment]
@mathieujobin actually - it isn't even that complex.
You can just edit the message right before the merge.

But that is something that the merger needs to remember to do.
Easier to have the subject reflect what is going to happen.

**Comment by mathieujobin** [comment]
> You can just edit the message right before the merge.

this is exactly what I was saying

**Comment by bootstraponline** [comment]
FYI [bitrise has shipped [ci skip]](http://devcenter.bitrise.io/tips-and-tricks/skip-a-build/) in PR titles and it's awesome. Would be nice for Travis to do the same.

**Comment by JustArchi** [comment]
I have another use case for this - I'm using crowdin for GitHub integration which automatically maintains ```l10n``` branch and opens pull request targetting ```master```, which I can merge.

I already excluded ```l10n``` branch from building in travis, but commits are fully automatic, meaning I don't have direct access to alter them. By marking this automatic PR with ```[ci skip]``` I'd stop trashing your servers with massive amount of commit build requests that do not make any sense, as every commit is small text change provided by upstream (so it can't even break the build, I'm building C# and strings are totally optional to be included or not).

So I'd also appreciate ```[ci skip]``` for PR title, as it'd benefit both you and me, but you more, as it doesn't matter that much for me whether those commits are queued up or not. Please 👍

**Comment by chambm** [comment]
I'd like to use this for skipping CI when making a pull request against another project's README.md. No reason to run tests for that...

**Comment by ehuelsmann** [comment]
Does 90 days of no activity mean that this feature is apparently no longer
desirable?!

>

**Comment by kbrock** [comment]
@ehuelsmann Nah. I know we still desire this.

From what I could tell, only the git commit messages are inspected, and the github title is not. Making this technically difficult since more data would need to flow down to travis.

Also, when it comes down to implementation, there are a few nuances that different users want implemented differently. Making it even less straightforward.

Thank you @joshk and travis team for considering it.

**Comment by disha94** [comment]
@mathieujobin this can be achieved in a different way too.
Travis provides these set of conditions for running conditional builds: https://docs.travis-ci.com/user/conditions-v1
You can use "commit-message" condition and then travis builds will only run if your commit has that regex present.
I have tried this out and it works as expected.

**Comment by bmitchinson** [comment]
Still would like to see this for work in progress PRs.

**Comment by iamareebjamal** [comment]
Maybe a PR label would be a better option

**Comment by cpettet** [comment]
My team would like to see this for WIP PRs.
